File: 09-backtracking/00046-Permutations/00046-permutations.cpp
// 46: Permutations
// https://leetcode.com/problems/permutations/



#include <iostream>
#include <vector>

using namespace std;

class Solution {
private:
    void backtrack(vector<int>& nums, int start, vector<vector<int>>& result) {
        if (start == nums.size()) {
            result.push_back(nums);
            return;
        }

        for (int i=start; i<nums.size(); i++) {
            swap(nums[i], nums[start]);
            backtrack(nums, start + 1, result);
            swap(nums[i], nums[start]);
        }
    }

public:
    // SOLUTOIN
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> result;
        backtrack(nums, 0, result);
        return result;
    }
};


int main() {
    Solution o;

    // INPUT
    vector<int> nums = {0,1};

    // OUTPUT
    auto result = o.permute(nums);

    cout << "["; for (auto x : result) {
        cout << "["; for (auto y : x) {
            cout << y << ",";
        } cout << "\b],";
    } cout << "\b]" << endl;

    return 0;
}

File: 09-backtracking/00046-Permutations/00046-permutations.py
# 46: Permutations
# https://leetcode.com/problems/permutations/


from typing import List

class Solution:
    # SOLUTION
    def permute(self, nums: List[int]) -> List[List[int]]:
        result = []

        def backtrack(start: int) -> None:
            if start == len(nums):
                result.append(nums[:])
                return

            for i in range(start , len(nums)):
                nums[i], nums[start] = nums[start], nums[i]
                backtrack(start+1)
                nums[i], nums[start] = nums[start], nums[i]

        backtrack(0)
        return result



if __name__ == "__main__":
    o = Solution()

    # INPUT
    nums = [0,1]

    # OUTPUT
    result = o.permute(nums)

    print(result)
